### PR TITLE
doc: address how to use TAEF

### LINKED ISSUE #1962: Docs should be updated to address how to use TAEF for the development of unit tests
In #1850 I was ask to add a test file to the unit tests.  
First hurdle was that `WEX`, which I found in the existing source files, is related to `TAEF`. Without knowing that, you are out of luck to find anything in the docs.  
We have TAEF.md in the doc folder but it tells nothing about how to use it for the development of unit tests and the link to the official documentation is broken (of course I was able to find it anyway).  
I was tearing my hair out when I tried to figure out why `WexTestClass.h` was not found even if it was included in the existing source files. Whenever I see a quoted file name like in `#include "WexTestClass.h"` I automatically assume that this file will be found in the project directory. But this is neither the case nor required, which should be mentioned.  
The official documentation already states that the WDK has to be installed as well as the Spectre mitigated libraries. Nevertheless I think it would be worth to leave some words and links about the preconditions in the docs here.

### PR DESCRIPTION
**Pull request by skyline75489**
## Summary of the Pull Request

This PR updates the doc in the repo to address how to use TAEF as part of the development workflow.

 
## PR Checklist
* [X] Closes #1962
* [X] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [X] Ironically, no need for tests.
* [ ] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [ ] Schema updated.
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Detailed Description of the Pull Request / Additional comments

Inspired by the original issue of @german-one and his comments in #1962, I try to add more contents to the documentation that hopefully will clarify the usage of TAEF. Personally I've stumbled upon several issues when it comes to TAEF when I first started working on PRs. I think this will help the community to learn TAEF and make contributions even more easily.

## Validation Steps Performed

Not really needed.

### CODE REVIEW

File: doc/TAEF.md
Comment (skyline75489) [inline doc/TAEF.md:5]: I gotta say I'm surprised that this file hasn't changed since "Initial release of the Windows Terminal source code". That's 17 months! This link to TAEF doc hasn't been working for god knows how long, and no one to my knowledge actually complaines about it?...

File: doc/TAEF.md
Comment (skyline75489) [inline doc/TAEF.md:15]: I don't fully know if this is true but I don't need to install the WDK for `te.exe` to be available. The `razzle,cmd` works just fine. Feel free to edit this PR.

**Comment by skyline75489** [comment]
Speaking of being helpful, I highly doubt that the WindowsTestPasses.md is going to help anyone outside the OS engineering team. Wonder why this was released in the first place.

File: CONTRIBUTING.md
Comment (zadjii-msft) [inline CONTRIBUTING.md:151]: _technically_, there is a way to run the tests in VS that @DHowett uses. I think it's more work than doing it from the commandline. I think it's fine to leave this as-is, just FYI

Comment (miniksa) [reply CONTRIBUTING.md:151]: The `Taef.Redist.Wlk` NuGet package has an MSTest/TAEF adapter inside of it that will let you browse and execute TAEF tests inside of Visual Studio. The performance of this mechanism leaves a lot to be desired and it is finnicky with the "debug this test" and other features you might expect when you see them light up inside Visual Studio. So it's probably best to just avoid it and do it the reliable way with `TE.exe` itself.

File: doc/TAEF.md
Comment (miniksa) [reply doc/TAEF.md:5]: I mean, the core contributors are always aware of how to run it without reading the doc. And the CI will crack anyone else over the head on their PR if they break a test. Also the top search engine result for something like 'TAEF usage' leads to the right place... so I can understand how it decayed. But glad you fixed it. :)

File: doc/TAEF.md
Comment (miniksa) [inline doc/TAEF.md:17]:
```suggestion
In a "normal" command line environment, `te.exe` may not be directly available. Try the following command to set up the development environment first:

```

File: doc/TAEF.md
Comment (miniksa) [inline doc/TAEF.md:21]: Do note that if you use PowerShell and `Import-Module .\tools\OpenConsole.psm1`, you will then be able to do `Invoke-OpenConsoleTests` (optionally with a few switches you can read inside that PowerShell script) and it will get things going for you as an alternative to the CMD way.

File: doc/TAEF.md
Comment (skyline75489) [reply doc/TAEF.md:5]: Well google helps a lot, there’s no doubt about it. But it does pose certain challenges to newcomers of this repo to realize how to even getting started with unit testing. It took me some time to even realize what is the keyword to google. I’m sure @german-one feels the same.

**Comment by miniksa** [comment]
> Speaking of being helpful, I highly doubt that the WindowsTestPasses.md is going to help anyone outside the OS engineering team. Wonder why this was released in the first place.

It's not going to help anyone external. But it also doesn't hurt anything to have it external.
The goal was more "if everything relevant to this project is in this one repo, it can't be lost." And I presumed that this repo would always have a core contributor team made of internal employees with some ability to follow the links in it.

I don't have an appropriate "internal only" place to keep this sort of information that isn't subject to decaying every few years for whatever reason. I don't think such a thing exists. The only thing that has stood the test of time is the source code repository. Docs sites decay, bug databases decay, shared folders decay.... but source code is forever.

File: CONTRIBUTING.md
Comment (j4james) [reply CONTRIBUTING.md:151]: FYI, when I need to debug tests in Visual Studio I set the debugging command on the test project to `$(OutDir)TE.EXE`, and the command arguments to `/inproc $(TargetPath)` (potentially with `/name` for a specific test case). I find that really useful if I need to step through the code. I'm not sure it works for all of the unit test projects, though.

File: doc/TAEF.md
Comment (german-one) [reply doc/TAEF.md:5]: Yes, that has been the main reason why I once filed the issue. If contributors are asked to write a unit test for their code for the first time they will likely get a problem to find the related documentation. If you look into existing unit tests you'll only find `WEX` rather than `TAEF`. Try to google WEX and you'll see how long it may take until you realize that TAEF is related to it and is the key to find the necessary information 😀

Comment (skyline75489) [reply doc/TAEF.md:5]: Speaking of search engine, I use Bing a lot. Like really a lot. So no hard feelings for anyone in the Bing team 😆

**Comment by skyline75489** [comment]
@miniksa Another thing that was bothering me is that it’s leaking some kind of internal information. Like the name OneCore or Osgwiki. But if it’s OK with you guys, I’m good.

File: CONTRIBUTING.md
Comment (skyline75489) [reply CONTRIBUTING.md]: @j4james That is a good approach I've also been using myself. But I think adding it here sounds like a overkill. It shouldn't be too hard for people to find the `/inproc` thing in TAEF documentation.

**Comment by skyline75489** [comment]
Thanks everyone for your comments. Please feel free to change anything that does not feel right before merging this, as I'm not a native speaker of this language.

**Comment by WSLUser** [comment]
Is it possible for https://github.com/microsoft/terminal/pull/7637 to get merged first so the linter can do it's thing on this markdown doc? I doubt it'll raise any warnings but at least we could know for sure the GH Action is working as it should.

**Comment by DHowett** [comment]
It is almost certainly not possible for something that doesn’t even have a solution proposed to be merged before something that does ;)

**Comment by DHowett** [comment]
(Sorry @WSLUser- you won’t get notifications for edited comments, and I edited mine.)

**Review by zadjii-msft — APPROVED** [review]

**Comment by skyline75489** [comment]
Pinging for the status of this PR. I think it's ready to be merged.

And by the way is the CI broken right now? It's not even running for some PRs...

File: doc/TAEF.md
Comment (DHowett) [inline doc/TAEF.md:17]:
```suggestion
> Note that you cannot easily run TAEF tests directly through Visual Studio. The `Taef.Redist.Wlk` NuGet package comes with an adapter that will let you browse and execute TAEF tests inside of Visual Studio, but its performance and reliability prevent us from recommending it here.
```

File: doc/TAEF.md
Comment (DHowett) [inline doc/TAEF.md:46]:
```suggestion
Import-Module .\tools\OpenConsole.psm1
Invoke-OpenConsoleTests
```

File: doc/TAEF.md
Comment (DHowett) [inline doc/TAEF.md:49]:
```suggestion
`Invoke-OpenConsoleTests` supports a number of options, which you can enumerate by running `Invoke-OpenConsoleTests -?`.
```

**Review by DHowett — CHANGES_REQUESTED** [review]

File: doc/TAEF.md
Comment (DHowett) [inline doc/TAEF.md:27]:
```suggestion
For the purposes of the OpenConsole project, you can run the tests using the `te.exe` that matches the architecture for which the test was built (x86/x64):
```

**Pull request merged**
